### PR TITLE
reference: clarify #[cfg] section

### LINKED ISSUE #39370: #[cfg(unknown_option = "...") should produce a warning
The rust reference https://doc.rust-lang.org/reference.html#conditional-compilation states:

> contains a string that can be checked against (#[cfg(bar = "baz")]). Currently, only compiler-defined configuration options can have the latter form.

Given that there's a fixed list of options, could rustc warn on invalid options? The following code compiles with no warnings:

```rust
#[cfg(i_dont_exist = "foo")]
extern crate bar;

fn main() {
}
```

This caught me out today when I wrote `target = "macos"` rather than `target_os = "macos"`.

### PR DESCRIPTION
**Pull request by durka**
Closes #39370, but maybe we (or clippy) should add a warning too.

### CODE REVIEW

**Comment by bluss** [comment]
It's still misleading, there is no "contains a string". Variables `feature="a"` and `feature="b"` can be independently set, so there is no `feature` and it doesn't contain either a or b.

**Comment by durka** [comment]
At least it does collapse whitespace though, so `--cfg 'x="y"'` and `--cfg 'x = "y"'` mean the same thing. So it does give the appearance of a property named `x` that contains the value `"y"`. However, `x` can take on multiple values at the same time.

**Comment by durka** [comment]
@bluss updated, what do you think?

File: src/doc/reference.md [hunk 1]
```diff
@@ -2114,10 +2114,19 @@ Sometimes one wants to have different compiler outputs from the same code,
...
+  either be defined by the compiler or specified on the command line, e.g.
+  `rustc --cfg foo`.
+- Key-value options, where a key may be associated with one or more values, e.g.
+  `#[cfg(bar = "baz")]`. These can also either be defined by the compiler or
+  specified on the command line, e.g. `rustc --cfg 'bar="baz"'` (whitespace around
+  the `=` is ignored).
+
+one that is either defined or not
+(`#[cfg(foo)]`), and the other that takes on one or more string values
+(`#[cfg(bar = "baz")]`).
```
Comment (bluss) [inline src/doc/reference.md]: better but why should we say there are two kinds of configuration options, when there is just one? `first` and `second="x"` are both just names of `cfg` vars. Also some old text left in here at the end.

**Comment by durka** [comment]
I think this might be a "compiler-colored glasses" kind of thing: it
certainly _feels_ like they are separate types of cfgs to the user.

**Comment by jmesmon** [comment]
As someone who just recently realized that "`first` and `second="x"` are both just names of cfg vars", my opinion is that knowing that is the actual meaning makes them much easier to think about.

Talking about associated keys seems to just confuse things.

**Comment by durka** [comment]
OK, new draft:

> Configuration options are boolean (on or off) and are named either with a single identifier (e.g. `foo`) or an identifier and a string (e.g. `foo = "bar"`; the quotes are required and spaces around the `=` are unimportant). Note that similar options, such as `foo`, `foo="bar"` and `foo="baz"` may each be set or unset independently.
>
> Configuration options are either provided by the compiler or passed in on the command line using `--cfg` (e.g. `rustc main.rs --cfg foo --cfg 'bar="baz"'`).

**Comment by steveklabnik** [comment]
durka: r=me with that change.

**Comment by durka** [comment]
OK, done. I also added a mention of `cfg!`. Will squash (wish you could do that with the web editor...).

**Comment by bluss** [comment]
@bors r=steveklabnik

Great!

**Pull request merged**